### ISSUE TITLE
twisted.web.http.Request does not cleanup temporary files when the quota limit is exceeded

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3176]
|<img alt="yasusii's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| yasusii reported|
|-|-|
|Trac ID|trac#3176|
|Type|defect|
|Created|2008-04-11 14:45:14Z|
|Branch|https://github.com/twisted/twisted/tree/request-content-cleanup-3176|

When exception is raised in handleContentChunk method, temporary file won't be removed unless you close the file.

Attachments:

* [http.py.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8de/8de5fc6c85883fbf7e8558aa49ea4ee16313e731/de671f4e1a1c7ad95bd2e9d52024877656afaa1b.diff) (462 bytes) - added by yasusii on 2008-04-11 14:45:33Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3176 3176
type__defect defect
reporter__yasusii yasusii
priority__normal normal
milestone__ 
branch__branches_request_content_cleanup_3176 branches/request-content-cleanup-3176
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1207925114000000 1207925114000000
changetime__1232037784000000 1232037784000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

The temporary file will eventually be garbage collected along with the rest of the request object.  However, that may take a while, so we should certainly try a bit harder here.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [26018]) Branching to 'request-content-cleanup-3176'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [26019]) Close the request body file in Request.connectionLost; also drop the reference to the HTTPChannel; change server.Request.connectionLost so it calls up to http.Request.connectionLost; tests for all

refs [#3176](https://github.com/twisted/twisted/issues/3176)
refs [#11181](https://github.com/twisted/twisted/issues/11181)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

This should be fixed.  As long as I was in the area, I also made Request drop its channel attribute.  This may break the cycle [#11181](https://github.com/twisted/twisted/issues/11181) is concerned with.  In the absence of any response from the original reporter of that ticket, I'm going to say that this branch will resolve both this ticket and [#11181](https://github.com/twisted/twisted/issues/11181).

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @exarkun|
|-|-|

Looks good to me.  Merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [26030]) Merge request-content-cleanup-3176

Author: exarkun
Reviewer: glyph
Fixes: [#11181](https://github.com/twisted/twisted/issues/11181)
Fixes: [#3176](https://github.com/twisted/twisted/issues/3176)

Explicitly close the file object containing the request body in the
request object's `connectionLost` method.  This frees the space
allocated for the request body as soon as the request object has
no further need for it, rather than waiting for the garbage
collector to get around to doing the same thing.

The reference the request holds to the HTTP protocol object is now
also broken in the `connectionLost` method to try to simplify the
garbage collector's job.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
